src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -176,60 +176,134 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_sigreq_complete(MPIR_Request ** sigreq)
+                                          MPI_Aint target_disp, int target_count,
+                                          MPI_Datatype target_datatype, MPIR_Win * win,
+                                          MPIDI_av_entry_t * addr, int rma_type,
+                                          MPIR_Request ** sigreq)
 {
     int mpi_errno = MPI_SUCCESS;
-    size_t o_size, t_size, alloc_iovs, alloc_iov_size;
-    MPIDI_OFI_win_request_t *req = NULL;
+    uint64_t flags;
+    struct fi_msg_rma msg;
+    struct fi_rma_iov riov;
+    struct iovec iov;
+    size_t target_bytes, origin_bytes;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_ALLOCATE_WIN_REQUEST_PUT_GET);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_ALLOCATE_WIN_REQUEST_PUT_GET);
+    MPIR_Datatype_get_size_macro(origin_datatype, origin_bytes);
+    origin_bytes *= origin_count;
+    MPIR_Datatype_get_size_macro(target_datatype, target_bytes);
+    target_bytes *= target_count;
 
-    o_size = sizeof(struct iovec);
-    t_size = sizeof(struct fi_rma_iov);
-    /* An upper bound on the iov limit */
-    MPIDI_OFI_count_iovecs(origin_count, target_count, 0, origin_datatype,
-                           target_datatype, MPI_DATATYPE_NULL, origin_bytes, target_bytes, 0,
-                           max_pipe, &alloc_iovs);
+    /* allocate request */
+    MPIDI_OFI_win_request_t *req = MPIDI_OFI_win_request_create();
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+    req->event_id = MPIDI_OFI_EVENT_ABORT;
+    req->next = MPIDI_OFI_WIN(win).syncQ;
+    MPIDI_OFI_WIN(win).syncQ = req;
 
-    alloc_iov_size = alloc_iovs * o_size + alloc_iovs * t_size;
+    /* allocate target iovecs */
+    struct iovec *target_iov;
+    void *target_base;
+    MPI_Aint total_target_iov_len;
+    MPI_Aint target_iov_offset = 0;
+    MPI_Aint max_iov_len;
+    MPI_Aint target_iov_len;
+    MPIR_Typerep_iov_len(target_count, target_datatype, target_bytes, &total_target_iov_len);
+    max_iov_len = MPL_MIN(total_target_iov_len, MPIR_CVAR_CH4_OFI_RMA_IOVEC_MAX);
+    target_iov = MPL_malloc(sizeof(struct iovec) * max_iov_len, MPL_MEM_RMA);
+    target_base =
+        (void *) (MPIDI_OFI_winfo_base(win, target_rank) +
+                  (target_disp * MPIDI_OFI_winfo_disp_unit(win, target_rank)));
+
+    /* allocate origin iovecs */
+    struct iovec *origin_iov;
+    MPI_Aint total_origin_iov_len;
+    MPI_Aint origin_iov_offset = 0;
+    MPI_Aint origin_iov_len;
+    MPIR_Typerep_iov_len(origin_count, origin_datatype, origin_bytes, &total_origin_iov_len);
+    max_iov_len = MPL_MIN(total_origin_iov_len, MPIR_CVAR_CH4_OFI_RMA_IOVEC_MAX);
+    origin_iov = MPL_malloc(sizeof(struct iovec) * max_iov_len, MPL_MEM_RMA);
+
+    MPIDI_OFI_INIT_SIGNAL_REQUEST(win, sigreq, &flags);
+    int i = 0, j = 0;
+    size_t msg_len;
+    while (i < origin_iov_len && j < target_iov_len) {
+        if (i == 0) {
```

Comment:
iovec population now happens in a single place. This loop is kind of complex, but at least it condenses the typerep calls now.